src/core/src/core_logic/PatchInstaller.py

```diff
@@ -736,3 +757,32 @@ def filter_out_excluded_updates(self, included_packages, included_package_versio
         self.composite_logger.log_debug(str(len(new_included_packages)) + " out of " + str(len(included_packages)) + " packages will remain included in the run.")
         return new_included_packages, new_included_package_versions
     # endregion
+
+    def get_max_batch_size(self, maintenance_window, package_manager):
+        """Returns maximum batch size for batch patching as per the time remaining in the maintenance window and time taken to install package by package manager"""
+        available_time_to_install_packages = maintenance_window.get_remaining_time_in_minutes()
+
+        if Constants.REBOOT_SETTINGS[self.execution_config.reboot_setting] != Constants.REBOOT_NEVER:
+            available_time_to_install_packages = available_time_to_install_packages - Constants.REBOOT_BUFFER_IN_MINUTES
+
+        available_time_to_install_packages = available_time_to_install_packages - Constants.BUFFER_TIME_FOR_BATCH_PATCHING_START_IN_MINUTES
+
+        self.composite_logger.log_debug("Avaliable time in minutes to install packages (after removing buffer time): {0}".format(available_time_to_install_packages))
+
+        max_batch_size_for_packages = 0
+
+        # Taking assumption that 1 of the packages in the batch takes maximum expected time to install and remaining packages take average time to install.
+        if available_time_to_install_packages > Constants.PACKAGE_INSTALL_EXPECTED_MAX_TIME_IN_MINUTES:
+            available_time_to_install_packages = available_time_to_install_packages - Constants.PACKAGE_INSTALL_EXPECTED_MAX_TIME_IN_MINUTES
+            max_batch_size_for_packages = max_batch_size_for_packages + 1
+
+            # Remaining packages take average expected time to install.
+            package_install_expected_avg_time_in_minutes = package_manager.get_package_install_expected_avg_time_in_minutes()
+            max_batch_size_for_packages += int(math.floor(available_time_to_install_packages / package_install_expected_avg_time_in_minutes))
+
+        if max_batch_size_for_packages > Constants.MAX_BATCH_SIZE_FOR_PACKAGES:
+            max_batch_size_for_packages = Constants.MAX_BATCH_SIZE_FOR_PACKAGES
```

Comment:
300 is batch size.
The packages will be installed in batches of 300.
There is loop in install_packages_in_batches function which try to install packages in batches of batch size.
If there are 350 packages, and batch size is 300 then there are two batches, 1 with size 300 and the other with size 50.

In case after first batch of 300 packages, there is not enough time to install 50 packages in a batch then second phase of batch patching will be done in which batch size is reduced to 1/10th. So, the new batch size would be 30 and packages would be installed in batch of 30.

In case there is not enough time to install 30 packages in a batch then we fallback to sequential patching in which packages are installed sequentially one by one.